app/grandchallenge/cases/models.py

```diff
@@ -961,16 +1008,38 @@ def start_dicom_import_job(self):
                 raise
 
     def _deidentify_files(self):
-        # this will need to iterate over all linked UserUploads
-        # for each, download the file from the uploads bucket, run the
-        # deidentifier and upload the deidentified file to the HI bucket
-        # for the time being just copy each upload to the HI bucket
+        deid = DicomDeidentifier(
+            # forced_inserts={
+            #     "StudyInstanceUID": self.study_instance_uuid,
+            #     "SeriesInstanceUID": self.series_instance_uuid,
+            #     "PatientID": PATIENT_ID,
+            #     "StudyID": STUDY_ID,
+            #     "StudyDate": STUDY_DATE,
+            #     "AccessionNumber": ACCESSION_NUMBER,
+            #     "SeriesNumber": SERIES_NUMBER,
+            # },
+            # assert_unique_value_for=["StudyInstanceUID", "SeriesInstanceUID"]
+        )
         for upload in self.user_uploads.all():
-            self._s3_client.copy(
-                CopySource={"Bucket": upload.bucket, "Key": upload.key},
-                Bucket=settings.AWS_HEALTH_IMAGING_BUCKET_NAME,
-                Key=f"{self._input_key}/{upload.filename}",
-            )
+            with (
+                SpooledTemporaryFile() as infile,
+                SpooledTemporaryFile() as outfile,
+            ):
+                self._s3_client.download_fileobj(
+                    Fileobj=infile,
+                    Bucket=upload.bucket,
+                    Key=upload.key,
+                )
+                infile.seek(0)
+
+                deid.deidentify_file(infile, output=outfile)
+
+                outfile.seek(0)
+                self._s3_client.upload_fileobj(
+                    Fileobj=outfile,
+                    Bucket=settings.AWS_HEALTH_IMAGING_BUCKET_NAME,
+                    Key=f"{self._input_key}/{upload.filename}",
+                )
```

Comment:
This set-up means that we would potentially upload one or multiple files in vain to the HI bucket if deidentification errors out on a subsequent file. The alternative would be to write all files to disk and wait for all to be deidentified before uploading.

Reply:
Oh I see. This is fine, the prefix can be wiped on error, same with removing the users uploads.